File: library/src/main/java/ke/tang/refresh/AnimationRefreshView.java
package ke.tang.refresh;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.ViewGroup;

import androidx.annotation.RawRes;

import com.airbnb.lottie.LottieAnimationView;
import com.airbnb.lottie.LottieComposition;
import com.airbnb.lottie.OnCompositionLoadedListener;

import java.io.InputStream;

/**
 * 支持Lottie动画的刷新View，可以用作Header或者Footer
 */
public class AnimationRefreshView extends LottieAnimationView implements RefreshLayout.Refreshable {
    private final static int STATUS_IDLE = 0;
    private final static int STATUS_DRAG = 1;
    private final static int STATUS_REFRESHING = 2;

    private int mStatus;

    private LottieComposition mRefreshComposition;
    private LottieComposition mPullComposition;

    public AnimationRefreshView(Context context) {
        this(context, null);
    }

    public AnimationRefreshView(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.animationRefreshHeaderViewStyle);
    }

    public AnimationRefreshView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.AnimationRefreshView, defStyleAttr, R.style.Widget_AnimationRefreshHeaderView);
        setPullAnimation(a.getResourceId(R.styleable.AnimationRefreshView_pullAnimation, 0));
        setRefreshAnimation(a.getResourceId(R.styleable.AnimationRefreshView_refreshAnimation, 0));
        a.recycle();
    }

    @Override
    public boolean isIndicator() {
        return false;
    }

    @Override
    public void onOffset(float delta) {
        if (mStatus == STATUS_IDLE) {
            setStatus(STATUS_DRAG);
        }

        if (mStatus == STATUS_DRAG) {
            setProgress(Math.min(1, Math.abs(delta)));
        }
    }

    @Override
    public void onRelease(boolean isTrigger) {
        if (isTrigger) {
            setStatus(STATUS_REFRESHING);
        }
    }

    @Override
    public void onReset() {
        setStatus(STATUS_IDLE);
    }

    private void setStatus(int status) {
        switch (status) {
            case STATUS_IDLE:
                loop(false);
                if (null != mPullComposition) {
                    setComposition(mPullComposition);
                }
                pauseAnimation();
                break;
            case STATUS_DRAG:
                break;
            case STATUS_REFRESHING:
                if (null != mRefreshComposition) {
                    setComposition(mRefreshComposition);
                }
                loop(true);
                playAnimation();
                break;
        }
        mStatus = status;
    }

    @Override
    public int getContentSize() {
        ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) getLayoutParams();
        return getHeight() + layoutParams.topMargin + layoutParams.bottomMargin;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        SavedState viewState = (SavedState) state;
        mStatus = viewState.mStatus;
        setStatus(mStatus);
        super.onRestoreInstanceState(viewState.getSuperState());
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState state = new SavedState(superState);
        state.mStatus = mStatus;
        return state;
    }

    /**
     * 设置拖拽时的动画
     *
     * @param animation
     */
    public void setPullAnimation(LottieComposition animation) {
        setPullAnimationInternal(animation);
    }

    void setPullAnimationInternal(LottieComposition animation) {
        mPullComposition = animation;
        setStatus(mStatus);
    }

    /**
     * 设置刷新时动画
     *
     * @param animation
     */
    public void setRefreshAnimation(LottieComposition animation) {
        setRefreshAnimationInternal(animation);
    }

    void setRefreshAnimationInternal(LottieComposition animation) {
        mRefreshComposition = animation;
        setStatus(mStatus);
    }

    void setPullAnimationStream(InputStream animation) {
        LottieComposition.Factory.fromInputStream(animation, new OnCompositionLoadedListener() {
            @Override
            public void onCompositionLoaded(LottieComposition composition) {
                setPullAnimationInternal(composition);
            }
        });
    }

    void setRefreshAnimationStream(InputStream animation) {
        LottieComposition.Factory.fromInputStream(animation, new OnCompositionLoadedListener() {
            @Override
            public void onCompositionLoaded(LottieComposition composition) {
                setRefreshAnimationInternal(composition);
            }
        });
    }

    /**
     * 设置拖拽时的动画
     *
     * @param res
     */
    public void setPullAnimation(@RawRes int res) {
        if (0 < res) {
            setPullAnimationStream(getResources().openRawResource(res));
        } else {
            setPullAnimationInternal(null);
        }
    }

    /**
     * 设置刷新时动画
     *
     * @param res
     */
    public void setRefreshAnimation(@RawRes int res) {
        if (0 < res) {
            setRefreshAnimationStream(getResources().openRawResource(res));
        } else {
            setRefreshAnimationInternal(null);
        }
    }

    public static class SavedState implements Parcelable {
        private int mStatus;
        private Parcelable superState;

        public SavedState(Parcelable superState) {
            this.superState = superState != BaseSavedState.EMPTY_STATE ? superState : null;
        }

        public Parcelable getSuperState() {
            return superState;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeInt(this.mStatus);
            dest.writeParcelable(this.superState, flags);
        }

        protected SavedState(Parcel in) {
            this.mStatus = in.readInt();
            this.superState = in.readParcelable(LottieAnimationView.class.getClassLoader());
        }

        public static final Creator<SavedState> CREATOR = new Creator<SavedState>() {
            @Override
            public SavedState createFromParcel(Parcel source) {
                return new SavedState(source);
            }

            @Override
            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }
}

File: library/src/main/java/ke/tang/refresh/AnimationRefreshView.kt
package ke.tang.refresh

import android.animation.ValueAnimator
import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IntegerRes
import androidx.annotation.RawRes
import com.airbnb.lottie.LottieAnimationView
import com.airbnb.lottie.LottieComposition
import com.airbnb.lottie.LottieCompositionFactory
import java.io.InputStream
import kotlin.math.abs
import kotlin.math.min

/**
 * Provide a refreshable view which contain lottie features
 * you can provide two animation resource
 * 1. refresh animation - a animation played when view in refreshing state, will play infinite until [RefreshLayout.completeRefresh] been invoked
 * 2. pull animation - a animation played when user pull, the animation process this controlled by pull distance
 */
class AnimationRefreshView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = R.attr.animationRefreshHeaderViewStyle,
        defStyleRes: Int = R.style.Widget_AnimationRefreshHeaderView
) : LottieAnimationView(context, attrs, defStyleAttr), Refreshable {
    private var mStatus: Int = STATUS_IDLE
        set(value) {
            when (value) {
                STATUS_IDLE -> {
                    if (null != mPullComposition) {
                        setComposition(mPullComposition!!)
                    }
                    repeatCount = 0
                    pauseAnimation()
                }
                STATUS_DRAG -> {
                }
                STATUS_REFRESHING -> {
                    if (null != mRefreshComposition) {
                        setComposition(mRefreshComposition!!)
                    }
                    repeatCount = ValueAnimator.INFINITE
                    playAnimation()
                }
            }
            field = value
        }

    private var mRefreshComposition: LottieComposition? = null
        set(value) {
            field = value
            mStatus = mStatus
        }
    private var mPullComposition: LottieComposition? = null
        set(value) {
            field = value
            mStatus = mStatus
        }

    init {
        with(context.obtainStyledAttributes(attrs, R.styleable.AnimationRefreshView, defStyleAttr, defStyleRes)) {
            setPullAnimation(getResourceId(R.styleable.AnimationRefreshView_pullAnimation, View.NO_ID))
            setRefreshAnimation(getResourceId(R.styleable.AnimationRefreshView_refreshAnimation, View.NO_ID))
            recycle()
        }
    }

    /**
     * Set a pull animation resource
     */
    fun setPullAnimation(@RawRes res: Int) {
        if (NO_ID != res) {
            setPullAnimationStream(resources.openRawResource(res), Integer.toHexString(res))
        } else {
            mPullComposition = null
        }
    }

    /**
     * Set a refresh animation resource
     */
    fun setRefreshAnimation(@RawRes res: Int) {
        if (NO_ID != res) {
            setRefreshAnimationStream(resources.openRawResource(res), Integer.toHexString(res))
        } else {
            mRefreshComposition = null
        }
    }

    private fun setPullAnimationStream(stream: InputStream, cacheKey: String? = null) {
        LottieCompositionFactory.fromJsonInputStream(stream, cacheKey)?.addListener { mPullComposition = it }
    }

    private fun setRefreshAnimationStream(stream: InputStream, cacheKey: String? = null) {
        LottieCompositionFactory.fromJsonInputStream(stream, cacheKey).addListener { mRefreshComposition = it }
    }

    override fun isIndicator() = false

    override fun onOffset(delta: Float) {
        if (mStatus == STATUS_IDLE) {
            mStatus = STATUS_DRAG
        }

        if (mStatus == STATUS_DRAG) {
            progress = min(1f, abs(delta))
        }
    }

    override fun onRelease(isTrigger: Boolean) {
        if (isTrigger) {
            mStatus = STATUS_REFRESHING
        }
    }

    override fun onReset() {
        mStatus = STATUS_IDLE
    }

    override fun getContentSize(isRefreshVertical: Boolean) = if (isRefreshVertical) {
        height + with(layoutParams as ViewGroup.MarginLayoutParams) { topMargin + bottomMargin }
    } else {
        width + with(layoutParams as ViewGroup.MarginLayoutParams) { leftMargin + rightMargin }
    }

    override fun onSaveInstanceState(): Parcelable? = SavedState(super.onSaveInstanceState()).apply {
        mStatus = this@AnimationRefreshView.mStatus
    }

    override fun onRestoreInstanceState(state: Parcelable?) = with(state as SavedState) {
        super.onRestoreInstanceState(superState)
        this@AnimationRefreshView.mStatus = mStatus
    }

    private class SavedState : BaseSavedState {
        var mStatus: Int = 0

        constructor(superState: Parcelable?) : super(superState)
        constructor(source: Parcel) : super(source) {
            with(source) {
                mStatus = source.readInt()
            }
        }

        override fun writeToParcel(out: Parcel, flags: Int) = with(out) {
            super.writeToParcel(out, flags)
            writeInt(mStatus)
        }

        companion object {
            @JvmField
            val CREATOR: Parcelable.Creator<SavedState> = object : Parcelable.Creator<SavedState> {
                override fun createFromParcel(source: Parcel) = SavedState(source)

                override fun newArray(size: Int): Array<SavedState?> = arrayOfNulls(size)
            }
        }
    }

    companion object {
        const val STATUS_IDLE = 0
        const val STATUS_DRAG = 1
        const val STATUS_REFRESHING = 2
    }
}